src/shims/time.rs

```diff
@@ -1,8 +1,99 @@
+
+/// A monotone clock used for `Instant` simulation.
+#[derive(Debug)]
+pub enum Clock {
+    Host {
+        /// The "time anchor" for this machine's monotone clock.
+        time_anchor: Instant,
+    },
+    Virtual {
+        /// The "current virtual time".
+        nanoseconds: AtomicU64,
+    },
+}
+
+impl Clock {
+    /// Create a new clock based on the availability of communication with the host.
+    pub fn new(communicate: bool) -> Self {
+        if communicate {
+            Self::Host { time_anchor: Instant::now() }
+        } else {
+            Self::Virtual { nanoseconds: 0.into() }
+        }
+    }
+
+    /// Get the current time relative to this clock.
+    pub fn get(&self) -> Duration {
+        match self {
+            Self::Host { time_anchor } => Instant::now().saturating_duration_since(*time_anchor),
+            Self::Virtual { nanoseconds } =>
+                Duration::from_nanos(nanoseconds.load(Ordering::Relaxed)),
+        }
+    }
+
+    /// Let the time pass for a small interval.
+    pub fn tick(&self) {
+        match self {
+            Self::Host { .. } => {
+                // Time will pass without us doing anything.
+            }
+            Self::Virtual { nanoseconds } => {
+                nanoseconds.fetch_add(NANOSECOND_PER_BASIC_BLOCK, Ordering::Relaxed);
+            }
+        }
+    }
+
+    /// Sleep for the desired duration.
+    pub fn sleep(&self, duration: Duration) {
+        match self {
+            Self::Host { .. } => std::thread::sleep(duration),
+            Self::Virtual { nanoseconds } => {
+                // Just pretend that we have slept for some time.
+                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
+            }
+        }
+    }
+
+    /// Compute `now + duration` relative to this clock.
+    pub fn checked_add_since_now(&self, duration: Duration) -> Option<Time> {
+        match self {
+            Self::Host { .. } => Instant::now().checked_add(duration).map(Time::Monotonic),
+            Self::Virtual { nanoseconds } =>
+                nanoseconds
+                    .load(Ordering::Relaxed)
+                    .checked_add(duration.as_nanos().try_into().unwrap())
+                    .map(|nanoseconds| Time::Virtual { nanoseconds }),
+        }
+    }
+
+    /// Compute `start + duration` relative to this clock where `start` is the instant of time when
+    /// this clock was created.
+    pub fn checked_add_since_start(&self, duration: Duration) -> Option<Time> {
+        match self {
+            Self::Host { time_anchor } => time_anchor.checked_add(duration).map(Time::Monotonic),
+            Self::Virtual { .. } =>
+                Some(Time::Virtual { nanoseconds: duration.as_nanos().try_into().unwrap() }),
+        }
+    }
+
+    /// Assert that this clock is a virtual one and get the current time in nanoseconds.
+    pub(crate) fn assert_virtual(&self) -> &AtomicU64 {
```

Comment:
This should not be public, it should be hidden inside this file (if it is needed at all).